scionlab/models/pki.py

```diff
@@ -0,0 +1,485 @@
+            -----END PRIVATE KEY-----
+        """).format(
+            algo=self.algorithm(),
+            ia=self.AS.isd_as_str(),
+            not_after=self._format_timestamp(self.not_after),
+            not_before=self._format_timestamp(self.not_before),
+            usage=self.usage,
+            version=self.version,
+            key=self.key
+        )
+
+    def algorithm(self):
+        """
+        Return the algorithm corresponding to this key, in the format expected for the .key file
+        """
+        if self.is_encryption_key():
+            return 'curve25519xsalsa20poly1305'
+        else:
+            return 'ed25519'
+
+    def is_encryption_key(self):
+        """
+        Returns True iff this is an encryption key. Otherwise, this is a signing key.
+        """
+        return self.usage == Key.DECRYPT
+
+    @staticmethod
+    def next_version(as_, usage):
+        prev = as_.keys.filter(usage=usage).aggregate(models.Max('version'))['version__max']
+        if prev:
+            return prev + 1
+        else:
+            return 1
+
+    @staticmethod
+    def _format_timestamp(dt):
+        """
+        The SCION key file format expects timestamps in a specific format:
+            2006-01-02 15:04:05-0700
+        """
+        assert dt.tzinfo is None, "Timestamps from DB are expected to be naive UTC datetimes"
+        return dt.replace(tzinfo=timezone.utc).strftime("%Y-%m-%d %H:%M:%S%z")
+
+    @staticmethod
+    def default_expiration(usage):
+        """
+        Return the default expiration time for keys of the given usage.
+        """
+        if usage in [Key.CERT_SIGNING,
+                     Key.TRC_ISSUING_GRANT,
+                     Key.TRC_VOTING_ONLINE,
+                     Key.TRC_VOTING_OFFLINE]:
+            return DEFAULT_EXPIRATION_CORE_KEYS
+        else:
+            return DEFAULT_EXPIRATION_AS_KEYS
+
+
+class TRCManager(models.Manager):
+    def create(self, isd):
+        """
+        Create a TRC for this ISD.
+
+        The version is incremented from the previous TRC. The voting offline keys related to the
+        previous TRC may be (precisely: for online key updates and sensitive updates) used to sign
+        the new TRC.
+
+        Requires at least one core AS in this ISD.
+
+        The latest keys for all core ASes are used. The validity period for the TRC is determined
+        based the validity of these keys.
+
+        :param isd ISD:
+        """
+
+        prev = isd.trcs.latest_or_none()
+
+        if prev:
+            version = prev.version + 1
+            prev_trc = prev.trc
+            prev_voting_offline = {key.as_id: _key_info(key)
+                                   for key in prev.voting_offline.all()}
+        else:
+            version = 1
+            prev_trc = None
+            prev_voting_offline = None
+
+        keys = {as_.as_id: _latest_core_keys(as_) for as_ in isd.ases.filter(is_core=True)}
+        all_keys = sum(keys.values(), [])
```

Comment:
Fun, hadn't heard of "Shlemiel the painter's algorithm" before

`core_ases = list(isd.ases.filter(is_core=True))`
`all_keys = [key for as_ in core_ases for key in _latest_core_keys(as_)]`

Reply:
Fixed, we even have a utils function for this. I like the sum thing though ;)